### DISCUSSION TITLE
ToolSettings.WithRawArgument

### DISCUSSION THREAD

**Discussion by jnm2** [cake-build/cake#3344] [Ideas]
Some tool settings use the fluent-mutation API style. This is great for one-liners but everything falls apart and you have to refactor the expression to multiple statements as soon as you need to reach for `ArgumentCustomization`, which will never be completely rare.

I'd like to propose the following extension method:

(`Raw` because it does no automatic escaping and so is susceptible to argument injection on Windows, a property it shares with most of Cake. 🙄)

```cake
public static T WithRawArgument<T>(this T settings, string rawArgument) where T : Cake.Core.Tooling.ToolSettings
{
    if (settings == null) throw new ArgumentNullException(nameof(settings));

    if (!string.IsNullOrEmpty(rawArgument))
    {
        var previousCustomizer = settings.ArgumentCustomization;
        if (previousCustomizer != null)
            settings.ArgumentCustomization = builder => previousCustomizer.Invoke(builder).Append(rawArgument);
        else
            settings.ArgumentCustomization = builder => builder.Append(rawArgument);
    }

    return settings;
}
```

Usage:

```cake
new MSBuildSettings()
    .WithTarget("Publish")
    .WithProperty("TargetFramework", "netcoreapp1.1")
    .WithProperty("NoBuild", "true")
    .WithRawArgument("/nologo")
```

**Comment by jnm2** [comment]
I wish we could tackle the bigger argument quoting/escaping injection issue across all of Cake, so maybe this isn't such a good idea yet.

**Comment by gep13** [comment]
>@jnm2 said...
I wish we could tackle the bigger argument quoting/escaping injection issue across all of Cake

Can you elaborate on what you mean by this?

**Comment by jnm2** [comment]
@gep13 Oh hey, looks like you're waiting for me on this one: https://github.com/cake-build/cake/issues/1272

**Comment by jnm2** [comment]
Also: https://github.com/cake-build/cake/issues/1898#issuecomment-339692650
